card/lib/card/stage_director.rb

```diff
@@ -0,0 +1,235 @@
+      catch_up_to_stage :prepare_to_store
+      run_single_stage :store, &block
+      run_single_stage :finalize
+    ensure
+      @from_trash = nil
+    end
+
+    # dirty marks are gone in this phase
+    def integration_phase
+      run_single_stage :integrate
+      run_single_stage :integrate_with_delay
+    ensure
+      unregister
+    end
+
+    def catch_up_to_stage next_stage
+      @stage ||= -1
+      (@stage + 1).upto(stage_index(next_stage)) do |i|
+        run_single_stage stage_symbol(i)
+      end
+    end
+
+    def call_after_store &block
+      @call_after_store ||= []
+      @call_after_store << block
+    end
+
+    def need_act
+      act_director = main_director
+      if !act_director
+        fail Card::Error, 'act requested without a main stage director'
+      end
+      act_director.act ||=  Card::Act.create(ip_address: Env.ip)
+      @card.current_act = @act = act_director.act
+    end
+
+    def main_director
+      if main?
+        self
+      else
+        DirectorRegister.act_director || (@parent && @parent.main_director)
+      end
+    end
+
+    def main?
+      @main
+    end
+
+    def running?
+      @running
+    end
+
+    private
+
+    def run_single_stage stage, &block
+      # puts "#{@card.name}: #{stage} stage".red
+      return if @card.errors.any?
+      @stage = stage_index stage
+
+      # in the store stage it can be necessary that
+      # other subcards must be saved before we save this card
+      if stage == :store
+        store &block
+      else
+        run_stage_callbacks stage
+        run_subdirector_stages stage unless stage == :finalize
+      end
+    rescue => e
+      @card.rescue_event e
+    end
+
+    def run_stage_callbacks stage
+      # we use abort :success in the :store stage for :save_draft
+      if stage_index(stage) <= stage_index(:store) && !main?
+        @card.abortable do
+          @card.run_callbacks :"#{stage}_stage"
+        end
+      else
+        @card.run_callbacks :"#{stage}_stage"
+      end
+    end
+
+    def run_subdirector_stages stage
+      @subdirectors.each do |subdir|
+        subdir.catch_up_to_stage stage
+      end
+    ensure
+      @card.handle_subcard_errors
+    end
+
+    # handle the store stage
+    # The tricky part here is to preserve the dirty marks on the subcards'
+    # attributes for the finalize stage.
+    # To achieve this we can't just call the :store and :finlaize callbacks on
+    # the subcards as  we do in the other phases.
+    # Instead we have to call `save` on the subcards
+    # and use the ActiveRecord :around_save callback to run the :store and
+    # :finalize stages
+    def store &save_block
+      if main? && !block_given?
+        fail Card::Error, 'need block to store main card'
+      end
+      # the block is the ActiveRecord block from the around save callback that
+      # saves the card
+      if block_given?
+        run_stage_callbacks :store
+        store_with_subcards &save_block
```

Comment:
Ambiguous block operator. Parenthesize the method arguments if it's surely a block operator, or add a whitespace to the right of the `&` if it should be a binary AND.